docs/spec/distributing.rst

```diff
@@ -10,72 +10,428 @@ Stub files
+file is found, the type checker should not read the corresponding "real" module.
+See :ref:`mro` for more information.
+
+.. _stub-file-supported-constructs:
+
+Supported Constructs
+^^^^^^^^^^^^^^^^^^^^
+
+This section lists constructs that type checkers will accept in stub files. Stub
+authors can safely use these constructs. If a construct is marked as
+"unspecified", type checkers may handle it as they best see fit or report an
+error. Linters should usually flag those constructs. Stub authors should avoid
+using them to ensure compatibility across type checkers.
+
+The `typeshed feature tracker <https://github.com/python/typeshed/labels/project%3A%20feature%20tracker>`_ tracks features from the ``typing`` module that are
+not yet supported by all major type checkers. Unless otherwise noted in this
+tracker, stub files support all features from the ``typing`` module of the
+latest released Python version. If a stub uses typing features from a later
+Python version than what the implementation supports, these features can be
+imported from ``typing_extensions`` instead of ``typing``
+
+For example, a stub could use ``Literal``, introduced in Python 3.8,
+for a library supporting Python 3.7+::
+
+    from typing_extensions import Literal
+
+    def foo(x: Literal[""]) -> int: ...
+
+Comments
+""""""""
+
+Standard Python comments are accepted everywhere Python syntax allows them.
+
+Two kinds of structured comments are accepted:
+
+* A ``# type: X`` comment at the end of a line that defines a variable,
+  declaring that the variable has type ``X``. However, :pep:`526`-style
+  variable annotations are preferred over type comments.
+* An error suppression comment at the end of any line. Common error suppression
+  formats are ``# type: ignore``, ``# type: ignore[error-class]``, and
+  ``# pyright: ignore[error-class]``. Type checkers may ignore error
+  suppressions that they don't support but should not error on them.
+
+Imports
+"""""""
+
+Stub files distinguish between imports that are re-exported and those
+that are only used internally. See :ref:`import-conventions`.
+
+Type aliases can be used to re-export an import under a different name::
+
+    from foo import bar as _bar
+    new_bar = _bar  # "bar" gets re-exported with the name "new_bar"
+
+Stubs support customizing star import semantics by defining a module-level
+variable called ``__all__``. In stubs, this must be a string list literal.
+Other types are not supported. Neither is the dynamic creation of this
+variable (for example by concatenation).
+
+When ``__all__`` is defined, exactly those names specified in ``__all__`` are
+imported::
+
+    __all__ = ['public_attr', '_private_looking_public_attr']
+
+    public_attr: int
+    _private_looking_public_attr: int
+    private_attr: int
+
+Type checkers support cyclic imports in stub files.
+
+Built-in Generics
+"""""""""""""""""
+
+:pep:`585` built-in generics are supported and should be used instead
```

Comment:
This is a good example of something that should be moved to a guide, not included in the typing spec.